Layers

yerlotic opened the pull request:
This PR introduces layers to achieve 3D workspace arrangement (for grid layout)

### Implementation details

Overview rendering (grid and linear) now accounts for layers. This layer is an int that determines workspaces being shown in overview.

Let's say we have a 4x3 grid. The overview shows only the first 12 workspaces.

After setting the layer to 1, the shown workspaces are [13-24]:

```
layer=0                      layer=1
1  2  3  4                   13  14  15  16
5  6  7  8  ⭢   layer=1   ⭢  17  18  19  20
9 10 11 12                   21  22  23  24
```

Layers are monitor-independent:
```
           Monitor 0
layer=0
*-----------*
|1   2  3  4|      13  14  15  16
|5   6  7  8|      17  18  19  20
|9  10 11 12|      21  22  23  24
*-----------*

          Monitor 1
layer=1
                  *--------------*
25 26 27 28       |37  38  39  40|
29 30 31 32       |41  42  43  44|
33 34 35 36       |45  46  47  48|
                  *--------------*
```

This PR introduces several new dispatchers:
 - `hyprtasking:setlayer` — changes currently visible workspaces by `rows*columns*argument`. Supports relative change
 - `hyprtasking:setlayerwindow` — changes currently visible workspaces the same as `hyprtasking:setlayer`. Moves window to the target workspace
 - `hyprtasking:move[window] in/out` switches to previous/next layer [and moves the active window]

And new options:
 - `hyprtasking:grid:layers` - limits the amount of layers for grid layout. Defaults to 1
 - `hyprtasking:grid:loop_layers` - should layers loop the same as workspaces? Defaults to true.

### Done

- [x] `hyprtasking:setlayerwindow` moves window to next layer (and in workspaces)
- [x] `hyprtasking:setlayer` doesn't allow switching less than zero and over maximum layer (specified by `hyprtasking:grid:layers`)
- [x] Documentation

Any feedback will be greatly appreciated

Closes #28

yerlotic committed:
Initial support for layers

yerlotic committed:
Add restrictions on layers, fix window movement

yerlotic committed:
Add `grid:loop_layers`

yerlotic committed:
Fix strange layer switching behaviour with `move_window`

yerlotic committed:
Update docs

yerlotic committed:
Rename to first_ws_offset

aaronjamt commented on src/main.cpp:
```diff
@@ -107,10 +107,113 @@ static SDispatchResult dispatch_move_window(std::string arg) {
     return {};
 }
 
+static void set_offset(int new_offset) {
+    for (PHTVIEW view : ht_manager->views) {
+        if (view == nullptr)
+            continue;
+        view->layout->first_ws_offset = new_offset;
+        Debug::log(
+            LOG,
+            "[Hyprtasking] offset was: {}, new: {}",
+            ht_manager->offset,
+            new_offset
+        );
+    }
+}
+
+static SDispatchResult dispatch_setoffset(std::string arg) {
+    if (ht_manager == nullptr)
+        return {.success = false, .error = "ht_manager is null"};
+
+    if (arg[0] == '+' || arg[0] == '-') {
+        ht_manager->offset += std::stoi(arg);
+    } else {
+        ht_manager->offset = std::stoi(arg);
+    }
+    set_offset(ht_manager->offset);
+    return {};
+}
```
Setting the offset directly doesn't update the currently focused workspace.

aaronjamt replied:
Correct, the issue is that if you change the offset, the currently visible workspace does not also change. Let's say you're using a 2x2 grid with 2 layers, if you're on the top-left workspace (0) and change the offset to 4, the current workspace is still 0, which isn't available. I would expect the current workspace to now be 4.

Maybe I should add that as a separate command?

aaronjamt replied:
> > Maybe I should add that as a separate command?
> 
> You mean a jump to a specific workspace?

I'm trying to make it so I can jump to a specific layer, rather than just going +/- relative to the current layer. 0bd61d3441549e6582e9efddca99402a93de2d84 works, so I can just do `hyprtasking:setoffset, 0` to go to the first layer, then `hyprtasking:setoffset, [grid size]` to go to the second layer, etc. Thanks for fixing that!

aaronjamt replied:
Can confirm, this works perfectly now! Thanks again!

yerlotic committed:
Move current active workspace on offset change

yerlotic committed:
Make dispatchers `setoffset` and `setlayer` behave the same way

yerlotic committed:
Per-monitor layers

aaronjamt commented on src/main.cpp:
```diff
@@ -108,6 +108,119 @@ static SDispatchResult dispatch_move_window(std::string arg) {
+    PHTVIEW view = ht_manager->get_view_from_cursor();
+    if (view == nullptr)
+        return;
+    Debug::log(
```
This PR works on Hyprland 0.53 if you change `Debug::log(` to `Log::logger->log(`

yerlotic replied:
Okay, this branch should now be up to date

yerlotic replied:
Well, [now](https://github.com/raybbian/hyprtasking/pull/60/commits/57676a37d2e03a9828ee936f02adf090a11d8b81) it is

yerlotic committed:
Fixes

yerlotic committed:
Chase hyprland v0.54

yerlotic committed:
Remove debug code

yerlotic committed:
Drag worky

yerlotic committed:
Use a specific `renderWindow` function

aaronjamt commented:
Just curious, is there anything blocking this PR? I've been daily-driving it since my first comment on Sep 15, 2025, 6 months ago, and haven't encountered any issues with it.

yerlotic committed:
Make the code better, fix focus loosing bug

yerlotic committed:
Fix layers for multiple monitors

yerlotic commented:
I think this PR is okay to merge now. I finally worked around this annoying bug when monitor would jump to a strange workspace and lose focus

[This](https://github.com/raybbian/hyprtasking/pull/60/changes#diff-34d21af3c614ea3cee120df276c9c4ae95053830d7f1d3deaf009a4625409ad2R130-R131) only drops the `hyprtasking:setlayer` dispatcher when closing

raybbian commented:
How is multi-monitor support? I havent had the time to test this but will in the next few days

yerlotic commented:
> How is multi-monitor support? I havent had the time to test this but will in the next few days

It looks like I just fixed it in [`cff9cc60`](https://github.com/raybbian/hyprtasking/pull/60/changes/cff9cc60e1495d99b1c7416588c503ba7fa89ea7)

I updated the PR description, it now shows how workspaces are distributed across layers in multi-monitor setup

yerlotic commented:
@raybbian, also maybe layers should just be another `hyprtasking:move[window]` argument like `in`/`out`, what do you think?
This might be a logical continuation of `move` dispatcher

Like this:

<img width="1920" height="1080" alt="layers" src="https://github.com/user-attachments/assets/44472e8c-8fb2-470c-bea4-410b0a04197d" />

yerlotic committed:
Add `hyprtasking:move[window] in/out` dispater

yerlotic commented:
@raybbian, this PR is ready for merging now!

raybbian commented:
Hi @yerlotic, thank you so much for your efforts! I'm quite happy with the large majority of everything here. I just wanted to ask what the workflow/use case for manually setting the offset would be? 

Given that the hierarchy goes monitors -> layers -> (row/col), is there a reason the workspace for a given position can't be something like 
```
view_id * LAYERS_PER_VIEW * ROWS_PER_LAYER * WORKSPACES_PER_ROW
+ layer_no * ROWS_PER_LAYER * WORKSPACES_PER_ROW
+ row * WORKSPACES_PER_ROW
+ col
```

yerlotic commented:
Hi @raybbian!

> Hi @yerlotic, thank you so much for your efforts! I'm quite happy with the large majority of everything here. I just wanted to ask what the workflow/use case for manually setting the offset would be?

@aaronjamt might have used it in the beginning (see [this comment](https://github.com/raybbian/hyprtasking/pull/60#issuecomment-3295721367)).
This also could be used for sliding the workspaces across the overview in linear and in grid modes (since layers are not supported in linear layout)

> Given that the hierarchy goes monitors -> layers -> (row/col), is there a reason the workspace for a given position can't be something like
> 
> ```
> view_id * LAYERS_PER_VIEW * ROWS_PER_LAYER * WORKSPACES_PER_ROW
> + layer_no * ROWS_PER_LAYER * WORKSPACES_PER_ROW
> + row * WORKSPACES_PER_ROW
> + col
> ```

So you suggest getting rid of `setoffset` entirely?
People have issues with multimonitor setups (#96) and this dispatcher might help them to do a more manual workspace configuration.
Also `setoffset` allows you to view windows from other monitors and this might help because hyprtasking currently doesn't move windows from dead/disconnected monitors back to active ones.

But overall I'm not against better software architecture (removing `setoffset`)

raybbian commented:
Hi @yerlotic! Sorry for the delay!

Regarding
> People have issues with multimonitor setups (https://github.com/raybbian/hyprtasking/issues/96) and this dispatcher might help them to do a more manual workspace configuration.

I will implement a dynamic workspace assignment for a new layout.

Regarding
> Also setoffset allows you to view windows from other monitors and this might help because hyprtasking currently doesn't move windows from dead/disconnected monitors back to active ones.

This is something I've faced as well and will add a dispatcher for.

I'm not the biggest fan of `setoffset`, I think there are better solutions to the problems. Will merge as soon as it is removed.

yerlotic commented:
> I'm not the biggest fan of setoffset, I think there are better solutions to the problems. Will merge as soon as it is removed.

Okay, got it

yerlotic committed:
No more offsets

yerlotic commented:
@raybbian Done!

yerlotic committed:
Add coordinate space
